dockstore-client/src/main/java/io/dockstore/client/cli/nested/BaseLanguageClient.java

```diff
@@ -0,0 +1,337 @@
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+import com.google.common.base.Joiner;
+import com.google.common.io.Files;
+import io.dockstore.client.cli.nested.NotificationsClients.NotificationsClient;
+import io.dockstore.common.Utilities;
+import io.swagger.client.ApiException;
+import io.swagger.client.model.ToolDescriptor;
+import org.apache.commons.configuration2.INIConfiguration;
+import org.apache.commons.exec.ExecuteException;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static io.dockstore.client.cli.ArgumentUtility.errorMessage;
+import static io.dockstore.client.cli.ArgumentUtility.exceptionMessage;
+import static io.dockstore.client.cli.ArgumentUtility.out;
+import static io.dockstore.client.cli.Client.API_ERROR;
+import static io.dockstore.client.cli.Client.ENTRY_NOT_FOUND;
+import static io.dockstore.client.cli.Client.GENERIC_ERROR;
+import static io.dockstore.client.cli.Client.IO_ERROR;
+
+/**
+ * A base class for all language clients
+ * Clients for CWL, WDL, Nextflow, etc should extend this and implement the abstract functions.
+ */
+public abstract class BaseLanguageClient {
+    private static final Logger LOG = LoggerFactory.getLogger(BaseLanguageClient.class);
+    protected final AbstractEntryClient abstractEntryClient;
+    protected INIConfiguration config;
+    protected String notificationsWebHookURL;
+    protected NotificationsClient notificationsClient;
+    protected BaseLauncher launcher;
+
+    // Fields set on initial launch
+    protected boolean isLocalEntry;
+    protected String entry;
+    protected String wdlOutputTarget;
+    protected String yamlParameterFile;
+    protected String jsonParameterFile;
+    protected String uuid;
+
+    // Fields generated during setup and running of entry
+    protected File tempLaunchDirectory;
+    protected File localPrimaryDescriptorFile;
+    protected File zippedEntryFile;
+    protected String selectedParameterFile;
+    protected File provisionedParameterFile;
+    protected String workingDirectory;
+    protected String stdout;
+    protected String stderr;
+
+    public BaseLanguageClient(AbstractEntryClient abstractEntryClient, BaseLauncher launcher) {
+        this.abstractEntryClient = abstractEntryClient;
+        this.launcher = launcher;
+    }
+
+    public void setLauncher(BaseLauncher launcher) {
+        this.launcher = launcher;
+    }
+
+    /**
+     * Selects the intended parameter file
+     */
+    public abstract String selectParameterFile();
+
+    /**
+     * Provision the input files based on the selected parameter file.
+     * Creates an updated version of the parameter file with new local file locations.
+     * Must set the variable workingDirectory
+     * @return Updated parameter file
+     */
+    public abstract File provisionInputFiles();
```

Comment:
In the case of WES I think we would not want to provision input files, i.e. download files specified with https, http, s3 or gs, to a local directory, but rather pass these URIs directly to the WES endpoint which would then have to do provisioning. The input files could be very large reference files and CRAM files so we could not effectively pass these via a POST command to the WES endpoint. I think I could pass a flag to provisionInputFiles that would indicate not to download files. The flag could be provided via the abstractEntryClient as is the localEntry flag. I haven't looked into enough detail in the wdlFileProvisioning.pullFiles and CWLClient pullFiles functions to see if I could just not call them. Thoughts?

Reply:
One approach would be to override the provisionInputFiles method and just return the parameter file that was passed to the launcher initially. Kind of like a stubbed function.

Reply:
Yes but the WDLClient, CWLClient would have to override the provisionInputFiles method right? However there will be a WESLauncher that won't be able to override the provisionInputFiles method because it is not a member of the 'launcher' class. So it would be helpful if the code that determines whether to download the input files (e.g. wdlFileProvisioning.pullFiles) is an abstract method in the BaseLauncher? Does that make sense? I.e. let the launcher decide whether to download the files to the local file system.

Reply:
You are right. I think a flag for something like skip-input-provisioning makes the most sense.

Reply:
OK